### DISCUSSION TITLE
Pre and post update functions / scripts

### DISCUSSION THREAD

**Discussion by pcace** [GameServerManagers/LinuxGSM#4492] [Ideas]
i am trying to find a way for pre and post update functions wich can be triggered whenever the game oder lgsm updates. 
Since i could not find anything alike yet, i thought it mitght be a good idea to open this here. 

Any ideas would be awesome!

**Comment by MicLieg** [comment]
At the moment this is not really feasible.
The only thing you could do would be to run cronjobs n minutes before or after "updates" at fixed times.

What are you trying to achieve?

**Comment by pcace** [comment]
Hi there, 

i am running a cs2 server with mods. one of these is metamod. Everytime cs2 updates, it also replaces a file called gameinfo.gi. in this file there is a line needed to enable metamod. 
So my thought was - whenever lgsm updates cs2 and finishes the update, it should also run a little script wich adds the line to the gameinfo.gi file, to make sure metamod is still running after update. 

Currently i am doing exactly what you said - i am more or less constantly checking via crontab if the desired line is in the file or not and if not its adding it. 

Cheers

**Comment by MicLieg** [comment]
There is a feature called Fixes: https://dev-docs.linuxgsm.com/technical/modules/fixes
These fixes are for creating files, seeds or setting options that the server needs but aren't handled during installation.

These fixes can be run a) after installation (usually once) or b) before each server startup: https://github.com/GameServerManagers/LinuxGSM/blob/master/lgsm/modules/fix.sh#L55

Possibly a cs2 metamod fix could be implemented using this feature.

However, as metamod isn't installed on every CS2 server, the fix should only be applied if metamod is installed, and perhaps only after an update, or if the "line" is missing.

**Reply by pcace** [reply]
Hi thanks a lot! 
so i have this script wich could be run each time (before) the server starts: 

addMetaMod.sh:
```bash
#!/bin/bash

# Remove Windows line endings (\r) from gameinfo.gi
sed -i 's/\r$//' serverfiles/game/csgo/gameinfo.gi

# Check if the line "Game    csgo/addons/metamod" exists in the file
if ! grep -q "Game    csgo/addons/metamod" serverfiles/game/csgo/gameinfo.gi; then

    # Open serverfiles/game/csgo/gameinfo.gi in the game/csgo directory
    sed -i 's/#.*\n//g' serverfiles/game/csgo/gameinfo.gi

    # Add Game csgo/addons/metamod to the SearchPaths section
    sed -i '/Game_LowViolence/{
        a             Game    csgo/addons/metamod
    }' serverfiles/game/csgo/gameinfo.gi

fi
```

so would it be enough to just cp it as fix_csgo.sh to lgsm/modules because it will be then fired up by the fix.sh? 

i dont quite get the manual here.

Thanks a gain for help!!

**Reply by pcace** [reply]
I did now add a cs2 fix to core_modules.sh: 

```bash

fix_cs2.sh() {
	modulefile="${FUNCNAME[0]}"
	fn_fetch_module
}
```
and added cs2 to pre start fixes in fixes.sh: 
```bash
apply_pre_start_fix=(arma3 armar ark av bt bo csgo cmw dst hw ins kf nmrih onset pvr rust rw samp sdtd sfc sof2 squad st tf2 terraria ts3 mcb mta unt vh wurm zmr cs2)
```
and also copied my script to /lgsm/modules/fix_cs2.sh

is that how it is intended to be used? or will this be replaced by any upcommung lgsm updates?

Cheers and thanks!


EDIT: Just looked at what happens if i update lgsm: 

```bash
checking GitHub module core_logs.sh...OK
checking GitHub module core_messages.sh...OK
checking GitHub module core_modules.sh...UPDATE
fetching GitHub core_modules.sh...OK
checking GitHub module core_steamcmd.sh...OK
checking GitHub module core_trap.sh...OK
checking GitHub module fix_cs2.sh...ERROR
removing module fix_cs2.sh...OK
checking GitHub module fix_csgo.sh...UPDATE
fetching GitHub fix_csgo.sh...OK
checking GitHub module fix.sh...UPDATE
fetching GitHub fix.sh...OK
checking GitHub module fix_steamcmd.sh...OK
```
meaning:
fix.sh and core_modules.sh get overwritten, changes are gone.

So how would i do it then?

Thanks!
